python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,346 @@
+    ClusterResourceConstraint,
+    GangResourceRequest,
+    NodeState,
+    ResourceRequest,
+    ResourceRequestByCount,
+)
+from ray.core.generated.instance_manager_pb2 import Instance
+
+# ============= Resource Scheduling Service API =======================
+#
+#  ResourceSchedulerService is a service that schedules resource bundles
+#  to nodes. It's used by the autoscaler to schedule resource bundles
+#  to determine the desired cluster size to satisfy the current resource
+#  demands.
+#
+
+
+@dataclass
+class NodeTypeConfig:
+    # Node type name
+    name: NodeType
+    # The minimal number of workers to be launched for this node type.
+    min_workers: int
+    # The maximal number of workers can be launched for this node type.
+    max_workers: int
+    # The resources on the node.
+    resources: Dict[str, float] = field(default_factory=dict)
+    # The labels on the node.
+    labels: Dict[str, str] = field(default_factory=dict)
+
+
+@dataclass
+class ResourceScheduleConfig:
+    # The node type configs.
+    node_type_configs: List[NodeTypeConfig] = field(default_factory=list)
+    # The max number of worker nodes to be launched for the entire cluster.
+    max_num_worker_nodes: Optional[int]
+
+
+@dataclass
+class ResourceScheduleRequest:
+    # TODO: This prob could be refactored into the ClusterStatus data class later.
+    # The current ray resource requests.
+    resource_requests: List[ResourceRequestByCount] = field(default_factory=list)
+    # The Gang resource requests.
+    gang_resource_requests: List[GangResourceRequest] = field(default_factory=list)
+    # cluster resource constraints.
+    cluster_resource_constraints: List[ClusterResourceConstraint] = field(
+        default_factory=list
+    )
+    # The ray nodes
+    node_states: List[NodeState] = field(default_factory=list)
+    # The current list of instances.
+    current_instances: List[Instance] = field(default_factory=list)
+    # The resource schedule config.
+    schedule_config: ResourceScheduleConfig
+
+
+@dataclass
+class ResourceScheduleReply:
+    # The infeasible resource bundles.
+    infeasible_resource_requests: List[ResourceRequestByCount] = field(
+        default_factory=list
+    )
+    # The infeasible gang resource bundles.
+    infeasible_gang_resource_requests: List[GangResourceRequest] = field(
+        default_factory=list
+    )
+    # The infeasible cluster resource constraints.
+    infeasible_cluster_resource_constraints: List[ClusterResourceConstraint] = field(
+        default_factory=list
+    )
+    # The target cluster shape, given the current resource demands and instances.
+    # Key is the node type name, value is the number of nodes.
+    # Note this might be "smaller" than the current cluster shape, since there
+    # could be resource requests constraints enforced.
+    target_cluster_shape: Dict[NodeType, int]
+    # The current cluster shape.
+    current_cluster_shape: Dict[NodeType, int]
```

Comment:
yeah, this could be omitted - i guess i will just return `nodes_to_launch` instead. 

The reason this is initially being returned is the routine to calculate the current cluster shape is now only done at the scheduler (callee). I think the caller could also do this.